File: examples/simple_demo.py
#!/usr/bin/env python3
"""
Simple example demonstrating the security frameworks mapping system.
"""

from mappings.system import mapping_system


def main():
    """Run the demo."""
    print("🔒 Security Frameworks Mapping System Demo\n")

    # 1. List all available frameworks
    print("📋 Available Frameworks:")
    frameworks = mapping_system.get_frameworks()
    for fw in frameworks:
        print(f"  • {fw.name} ({fw.id}) - {len(fw.get_all_controls())} controls")
    print()

    # 2. Analyze gaps between SOC 2 and ISO 27001
    print("🔍 Gap Analysis: SOC 2 → ISO 27001")
    soc2_to_iso = mapping_system.analyze_compliance("soc2", "iso27001")
    print(f"  Coverage: {soc2_to_iso.coverage_percentage:.1f}%")
    print(f"  Mapped Controls: {soc2_to_iso.mapped_controls}/{soc2_to_iso.total_source_controls}")
    print(f"  Unmapped: {len(soc2_to_iso.gaps['source'])} SOC 2 controls need additional work")
    print()

    # 3. Find mappings for a specific control
    print("🔗 Mappings for SOC 2 CC6.1 (Access Controls):")
    cc61_mappings = mapping_system.get_engine().find_mappings_for_control("soc2", "CC6.1")
    for mapping in cc61_mappings:
        target_fw = mapping.target_framework if mapping.source_framework == "soc2" else mapping.source_framework
        target_ctrl = mapping.target_control if mapping.source_framework == "soc2" else mapping.source_control
        print(f"  → {target_fw.upper()} {target_ctrl} [{mapping.mapping_type}, {mapping.confidence * 100:.0f}% confidence]")
    print()

    # 4. Generate compliance matrix
    print("📊 Compliance Coverage Matrix:")
    matrix = mapping_system.generate_compliance_matrix()
    framework_ids = matrix.frameworks

    # Print header
    header = "Source\\Target\t" + "".join(f"{fid:10}" for fid in framework_ids)
    print(header)

    # Print rows  
    for source in framework_ids:
        row = f"{source:15}"
        for target in framework_ids:
            if source == target:
                row += "    -     "
            else:
                coverage = matrix.matrix[source][target]["coverage"]
                row += f"{coverage:6.1f}%   "
        print(row)
    print()

    # 5. Find similar controls
    print("🔍 Similar Controls to SOC 2 CC6.1:")
    similar = mapping_system.find_potential_mappings("soc2", "CC6.1", 0.6)
    for match in similar[:3]:
        print(f"  • {match['framework_id'].upper()} {match['control_id']}: {match['similarity'] * 100:.0f}% similar")
        print(f"    Suggested mapping: {match['suggested_mapping_type']}")
    print()

    # 6. Identify critical gaps
    print("🚨 Critical Security Gaps to Address:")
    for source_id in ["soc2", "iso27001", "nist-csf"]:
        for target_id in ["soc2", "iso27001", "nist-csf"]:
            if source_id != target_id:
                analysis = mapping_system.analyze_compliance(source_id, target_id)
                critical_gaps = [gap for gap in analysis.gaps["target"] 
                               if gap["risk_level"] in ["critical", "high"]]
                
                if critical_gaps:
                    print(f"  {source_id.upper()} → {target_id.upper()}: {len(critical_gaps)} critical gaps")
                    for gap in critical_gaps[:2]:
                        print(f"    • {gap['id']}: {gap['title']}")
    print()

    # 7. Export data for further analysis
    print("💾 Exporting mapping data...")
    export_data = mapping_system.export()
    print(f"  Frameworks: {len(export_data['frameworks'])}")
    print(f"  Mappings: {len(export_data['mappings'])}")
    verified_mappings = len([m for m in export_data['mappings'] if m['verified']])
    print(f"  Verified mappings: {verified_mappings}")
    print()

    print("✅ Demo complete! Use the CLI tools for more detailed analysis:")
    print("  python -m mappings frameworks")
    print("  python -m mappings gaps soc2 iso27001")
    print("  python -m mappings report summary")


if __name__ == "__main__":
    main()

File: examples/simple-demo.js
import { mappingSystem } from '../src/index.js';

/**
 * Simple example demonstrating the security frameworks mapping system
 */

console.log('🔒 Security Frameworks Mapping System Demo\n');

// 1. List all available frameworks
console.log('📋 Available Frameworks:');
const frameworks = mappingSystem.getFrameworks();
frameworks.forEach(fw => {
  console.log(`  • ${fw.name} (${fw.id}) - ${fw.getAllControls().length} controls`);
});
console.log();

// 2. Analyze gaps between SOC 2 and ISO 27001
console.log('🔍 Gap Analysis: SOC 2 → ISO 27001');
const soc2ToIso = mappingSystem.analyzeCompliance('soc2', 'iso27001');
console.log(`  Coverage: ${soc2ToIso.coveragePercentage.toFixed(1)}%`);
console.log(`  Mapped Controls: ${soc2ToIso.mappedControls}/${soc2ToIso.totalSourceControls}`);
console.log(`  Unmapped: ${soc2ToIso.gaps.source.length} SOC 2 controls need additional work`);
console.log();

// 3. Find mappings for a specific control
console.log('🔗 Mappings for SOC 2 CC6.1 (Access Controls):');
const cc61Mappings = mappingSystem.getEngine().findMappingsForControl('soc2', 'CC6.1');
cc61Mappings.forEach(mapping => {
  const targetFw = mapping.targetFramework === 'soc2' ? mapping.sourceFramework : mapping.targetFramework;
  const targetCtrl = mapping.targetFramework === 'soc2' ? mapping.sourceControl : mapping.targetControl;
  console.log(`  → ${targetFw.toUpperCase()} ${targetCtrl} [${mapping.mappingType}, ${(mapping.confidence * 100).toFixed(0)}% confidence]`);
});
console.log();

// 4. Generate compliance matrix
console.log('📊 Compliance Coverage Matrix:');
const matrix = mappingSystem.generateComplianceMatrix();
const frameworkIds = Object.keys(matrix);

// Print header
console.log('Source\\Target\t', frameworkIds.map(id => id.padEnd(10)).join(' '));

// Print rows  
frameworkIds.forEach(source => {
  let row = source.padEnd(15);
  frameworkIds.forEach(target => {
    if (source === target) {
      row += '    -     ';
    } else {
      const coverage = matrix[source][target]?.coverage || 0;
      row += `${coverage.toFixed(1)}%`.padEnd(10);
    }
  });
  console.log(row);
});
console.log();

// 5. Find similar controls
console.log('🔍 Similar Controls to SOC 2 CC6.1:');
const similar = mappingSystem.findPotentialMappings('soc2', 'CC6.1', 0.6);
similar.slice(0, 3).forEach(match => {
  console.log(`  • ${match.frameworkId.toUpperCase()} ${match.controlId}: ${(match.similarity * 100).toFixed(0)}% similar`);
  console.log(`    Suggested mapping: ${match.suggestedMappingType}`);
});
console.log();

// 6. Identify critical gaps
console.log('🚨 Critical Security Gaps to Address:');
['soc2', 'iso27001', 'nist-csf'].forEach(sourceId => {
  ['soc2', 'iso27001', 'nist-csf'].forEach(targetId => {
    if (sourceId !== targetId) {
      const analysis = mappingSystem.analyzeCompliance(sourceId, targetId);
      const criticalGaps = analysis.gaps.target.filter(gap => 
        gap.riskLevel === 'critical' || gap.riskLevel === 'high'
      );
      
      if (criticalGaps.length > 0) {
        console.log(`  ${sourceId.toUpperCase()} → ${targetId.toUpperCase()}: ${criticalGaps.length} critical gaps`);
        criticalGaps.slice(0, 2).forEach(gap => {
          console.log(`    • ${gap.id}: ${gap.title}`);
        });
      }
    }
  });
});
console.log();

// 7. Export data for further analysis
console.log('💾 Exporting mapping data...');
const exportData = mappingSystem.export();
console.log(`  Frameworks: ${exportData.frameworks.length}`);
console.log(`  Mappings: ${exportData.mappings.length}`);
console.log(`  Verified mappings: ${exportData.mappings.filter(m => m.verified).length}`);
console.log();

console.log('✅ Demo complete! Use the CLI tools for more detailed analysis:');
console.log('  npm run analyze frameworks');
console.log('  npm run analyze gaps soc2 iso27001');
console.log('  npm run report summary');